### PR TITLE
logging: Add string validation to detect %p misuse

### PR DESCRIPTION
**Pull request by nordic-krch**
Logging shall not use character pointers with %p because in certain configurations it may lead to memory faults. A compile time detection is added. If faulty usage is detected then message is replaced with error message which indicates which message failed and what shall be done (casting to a pointer of different type).

Validation is enabled only for configurations which remove strings from binary as otherwise it may impact CI execution time.

Validation macros are added to cbprintf.

`Z_CBPRINTF_NONE_CHAR_PTR_COUNT` counts number of none character pointers in the arguments list.
`Z_CBPRINTF_P_COUNT` counts number of %p occurrences in the format string. It relies on the assumption that string is fixed so result is known at compile time.

If results of both macros are equal it means that string is ok.

### CODE REVIEW

**Review by dcpleung — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]
Not going to block. This is a bit mind-boggling tbh, but it sounds as though it fixes a bug.

**Comment by cfriedt** [comment]
This might be related to the recent upgrade to PicoLibC 1.8.5. I'm seeing strange things happening with constant strings elsewhere after the upgrade. Almost like things are optimized out.

cc @henrikbrixandersen

File: subsys/logging/Kconfig.misc [hunk 1]
```diff
@@ -70,6 +70,15 @@ config LOG_FMT_SECTION_STRIP
...
 	imply LINKER_DEVNULL_MEMORY
+	imply LOG_FMT_STRING_VALIDATE
+
+config LOG_FMT_STRING_VALIDATE
+	bool "Validate logging strings"
+	help
+	  Logging strings cannot use %p with character pointers. They shall be
+	  casted to the pointer of other type. When feature is enabled, at
+	  compile time preprocessor detects strings with %p without casting and
+	  report error at runtime.
```
Comment (pdgendt) [inline subsys/logging/Kconfig.misc:81]: Some description improvements. What is the "other type" you are referring to?

```suggestion
	  Logging strings cannot use %p with character pointers. They shall be
	  casted to the pointer of another type. With this feature is enabled, at
	  compile time, the preprocessor detects strings with %p without casting
	  and reports an error at runtime.
```

File: include/zephyr/sys/cbprintf_internal.h [hunk 2]
```diff
@@ -138,6 +138,356 @@ extern "C" {
...
+	((sizeof(fmt) >= 249 && fmt[246] != '%' && fmt[247] == '%' && fmt[248] == 'p') ? 1 : 0) + \
+	((sizeof(fmt) >= 250 && fmt[247] != '%' && fmt[248] == '%' && fmt[249] == 'p') ? 1 : 0) + \
+	((sizeof(fmt) >= 251 && fmt[248] != '%' && fmt[249] == '%' && fmt[250] == 'p') ? 1 : 0) + \
+	((sizeof(fmt) >= 252 && fmt[249] != '%' && fmt[250] == '%' && fmt[251] == 'p') ? 1 : 0) + \
+	((sizeof(fmt) >= 253 && fmt[250] != '%' && fmt[251] == '%' && fmt[252] == 'p') ? 1 : 0) + \
+	((sizeof(fmt) >= 254 && fmt[251] != '%' && fmt[252] == '%' && fmt[253] == 'p') ? 1 : 0) + \
+	((sizeof(fmt) >= 255 && fmt[252] != '%' && fmt[253] == '%' && fmt[254] == 'p') ? 1 : 0) + \
+	((sizeof(fmt) >= 256 && fmt[253] != '%' && fmt[254] == '%' && fmt[255] == 'p') ? 1 : 0)
+
+/** @brief Determine if all %p arguments has none character pointer argument.
```
Comment (pdgendt) [inline include/zephyr/sys/cbprintf_internal.h:470]:
```suggestion
/** @brief Determine if all %p arguments are none character pointer arguments.
```

File: include/zephyr/sys/cbprintf_internal.h [hunk 3]
```diff
@@ -138,6 +138,356 @@ extern "C" {
...
+ *
+ * Static package creation relies on the assumption that character pointer is
+ * using only for %s argument. To not confuse it with %p, any character pointer
+ * that is using for %p should be casted to a pointer of different type, e.g.
+ * void *. This macro can be used to determine at compile time if such casting
+ * is missing. It is determined at compile time but cannot be used for static
+ * assert so only runtime error reporting can be added.
```
Comment (pdgendt) [inline include/zephyr/sys/cbprintf_internal.h:477]:
```suggestion
 * Static package creation relies on the assumption that character pointers are
 * only using %s arguments. To not confuse it with %p, any character pointer
 * that is used with %p should be casted to a pointer of a different type, e.g.
 * void *. This macro can be used to determine, at compile time, if such casting
 * is missing. It is determined at compile time but cannot be used for static
 * assertion so only runtime error reporting can be added.
```

**Review by pdgendt — COMMENTED** [review]
Some documentation improvements.

File: subsys/logging/Kconfig.misc (hunk 1 above)
Comment (nordic-krch) [reply subsys/logging/Kconfig.misc]: Added `another type (e.g. void *)`. Basically it means that it shall be casted to anything other than `char *` or `unsigned char *` and simplest is `void *` to indicate that it is a pointer and for `%p` it does not matter what is the type.

**Review by pdgendt — DISMISSED** [review]
Minor nit

File: subsys/logging/Kconfig.misc [hunk 4]
```diff
@@ -70,6 +70,15 @@ config LOG_FMT_SECTION_STRIP
...
 	depends on LINKER_DEVNULL_SUPPORT
 	imply LINKER_DEVNULL_MEMORY
+	imply LOG_FMT_STRING_VALIDATE
+
+config LOG_FMT_STRING_VALIDATE
+	bool "Validate logging strings"
+	help
+	  Logging strings cannot use %p with character pointers. They should be
+	  casted to the pointer of another type (e.g. void *). With this feature
+	  is enabled, at compile time, the preprocessor detects strings with %p
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply subsys/logging/Kconfig.misc]: done

**Review by pdgendt — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

File: include/zephyr/logging/log_core.h [hunk 5]
```diff
@@ -184,12 +184,37 @@ static inline char z_log_minimal_level_to_char(int level)
 	case LOG_LEVEL_DBG:
 		return 'D';
 	default:
+
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply include/zephyr/logging/log_core.h]: done

**Comment by nordic-krch** [comment]
@pdgendt @dcpleung can you re-approve. Just removed blank line noticed by @cfriedt.

**Review by pdgendt — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**